gql/dsl.py

```diff
@@ -1,179 +1,378 @@
+            :code:`ds.Query.human.args(id=1000)`
+
+        :param \**kwargs: the arguments (keyword=value)
+        :return: itself
+
+        :raises KeyError: if any of the provided arguments does not exist
+                          for this field.
+        """
+
         added_args = []
         for name, value in kwargs.items():
             arg = self.field.args.get(name)
             if not arg:
                 raise KeyError(f"Argument {name} does not exist in {self.field}.")
-            arg_type_serializer = get_arg_serializer(arg.type, known_serializers=dict())
+            arg_type_serializer = self._get_arg_serializer(arg.type)
             serialized_value = arg_type_serializer(value)
             added_args.append(
                 ArgumentNode(name=NameNode(value=name), value=serialized_value)
             )
-        ast_field = self.ast_field
-        ast_field.arguments = FrozenList(ast_field.arguments + added_args)
+        self.ast_field.arguments = FrozenList(self.ast_field.arguments + added_args)
+        log.debug(f"Added arguments {kwargs} in field {self!r})")
         return self
 
-    @property
-    def ast(self):
-        return self.ast_field
-
-    def __str__(self):
-        return print_ast(self.ast_field)
-
-
-def selection_field(field):
-    if isinstance(field, DSLField):
-        return field
-
-    raise TypeError(f'Received incompatible query field: "{field}".')
+    def _get_arg_serializer(self, arg_type: GraphQLInputType,) -> Serializer:
+        if isinstance(arg_type, GraphQLNonNull):
+            return self._get_arg_serializer(arg_type.of_type)
+        elif isinstance(arg_type, GraphQLInputField):
+            return self._get_arg_serializer(arg_type.type)
+        elif isinstance(arg_type, GraphQLInputObjectType):
+            if arg_type in self.known_arg_serializers:
+                return cast(Serializer, self.known_arg_serializers[arg_type])
+            self.known_arg_serializers[arg_type] = None
+            serializers = {
+                k: self._get_arg_serializer(v) for k, v in arg_type.fields.items()
+            }
+            self.known_arg_serializers[arg_type] = lambda value: ObjectValueNode(
+                fields=FrozenList(
+                    ObjectFieldNode(name=NameNode(value=k), value=serializers[k](v))
+                    for k, v in value.items()
+                )
+            )
+            return cast(Serializer, self.known_arg_serializers[arg_type])
+        elif isinstance(arg_type, GraphQLList):
+            inner_serializer = self._get_arg_serializer(arg_type.of_type)
+            return lambda list_values: ListValueNode(
+                values=FrozenList(inner_serializer(v) for v in list_values)
+            )
+        elif isinstance(arg_type, GraphQLEnumType):
+            return lambda value: EnumValueNode(
+                value=cast(GraphQLEnumType, arg_type).serialize(value)
+            )
 
+        return lambda value: ast_from_value(
```

Comment:
Maybe similar as above: `elif isinstance(arg_type, GraphQLScalarType): ... else: throw error`

Reply:
The problem is that it is impossible in theory that the error happens if `_get_arg_serializer` is called by the `args` method.
So it means we have to add a meaningless test for this in order for the raise error line to get covered by the tests.
I propose instead to use `assert isinstance(arg_type, GraphQLScalarType)` here.

Reply:
Right, an assert may be appropriate in such a situation where you have a condition that should theoretically *never* fail. Note that it will be stripped away with "-O" which can be seen as an advantage (better performance in production) or disadvantage ("impossible" but still happening errors in production not detected). If you use asserts, then you should do it consistently for such cases in the whole codebase, (see also [here](https://github.com/graphql-python/graphql-core-legacy/issues/244)).

One of the reasons to add an "if isinstance" or an "assert" is that mypy should (in theory) understand that so that you don't need the "cast" (not sure if that works in the latest version of mypy). It is also a hint to the reader of the code. Btw, GraphQL-Core also has an `assert_scalar_type` function that you can use here. It raises a TypeError though instead of an AssertionError. Its return value is GraphQLScalarType so that should also help mypy.

I also add a "# pragma: no cover" to not decrease coverage for such code that should in theory never execute.

Reply:
See my last commit, this method was removed so this problem as well :laughing: